2015/presentations/Business logic with PostgreSQL and Python/text.md

```diff
@@ -0,0 +1,399 @@
+
+Also on some systems if you try to run Python and you’re getting this kind of error
+
+    py ➤ bin/python2.7
+
+**bin/python2.7: error while loading shared libraries: libpython2.7.so.1.0: cannot open shared object file: No such file or directory**
+
+Then you need to add lib path of your newly compiled python to the system lib path
+
+    export LD_LIBRARY_PATH=$LD_LIBRARY_PATH:/opt/py/lib
+    
+## PostgreSQL with Python support
+
+Download source code.
+
+    wget https://ftp.postgresql.org/pub/source/v9.4.4/postgresql-9.4.4.tar.bz2
+
+Compile with Python support from */opt/py* directory
+
+    ./configure –prefix=/opt/pgsql —with-python PYTHON=/opt/py/bin/python
+    make
+    make install
+
+Now compile and install PostgreSQL plugins
+
+    cd ~/stuff/postgresql-9.4.3/contrib
+    make
+    make install
+
+## Create DB
+
+Now it is time to start PostgreSQL. First you have to initialize name space.
+
+    mkdir /opt/pg_data
+    /opt/pgsql/bin/initdb /opt/pg_data
+
+and start DB…
+
+    /opt/pgsql/bin/postmaster -p 5432 -D /opt/pg_data/
+
+Create new DB.
+
+    /opt/pgsql/bin/createdb -h localhost  -E utf8 pie
+
+Please remember about the correct encoding for DB! Now create new language for database pie
+
+    /opt/pgsql/bin/createlang -h localhost  -d pie plpythonu
+    
+And that's it. Simple as that. Your new DB has full Python support and we can start organizing business logic in there.
+
+## Word of the day
+
+Before you are going to create your first plPython (PostgreSQL stored Python functions) you have to know how it works. Your newly compiled Python and its modules are fully accessible from plPython. This means that the entire Python standard library is fully accessible when writing your business logic. Also you can install any kind of module which you may use later from your functions.
+
+I have to make some other clarification here. Term "procedural" and "function" can be a little bit confusing here. Procedural programming is not the only pattern that you can use in plPython. Once you create plpy function then the actual body of it can be pretty complex object. For sure I am not trying to convince you here that object programming in such a case like business logic in DB is the right way. Zen of Python described it as below. Read it again, again, and... again before you start writing actually any Python code.
+
+> Beautiful is better than ugly.
+
+> Explicit is better than implicit.
+
+> Simple is better than complex.
+
+> Flat is better than nested.
+
+> Readability counts.
+
+> If the implementation is hard to explain, it’s a bad idea
+
+What I am trying to say here is that just to be careful with creating plpy functions. Not to heavy, not too complex. Simplicity can bring you a lot of benefits later when you have to squash a bug.

```

Comment:
What I am trying to point out here is that you should be careful when creating plpy functions, they should not be too complex. Simplicity can bring you a lot of benefits, especially when debugging the code to nail a bug.